File: port_js/org/kbinani/vsq/TextStream.js
/*
 * TextStream.js
 * Copyright (C) 2008-2010 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
if( org == undefined ) var org = {};
if( org.kbinani == undefined ) org.kbinani = {};
if( org.kbinani.vsq == undefined ) org.kbinani.vsq = {};
if( org.kbinani.vsq.TextStream == undefined ){

    org.kbinani.vsq.TextStream = function(){
        this.array = new Array();
        this.length = 0;
        this.position = -1;
    };

    org.kbinani.vsq.TextStream.prototype = {
        /**
         * @return [int]
         */
        getPointer : function() {
            return this.position;
        },

        /**
         * @param value [int]
         * @return [void]
         */
        setPointer : function( value ) {
            this.position = value;
        },

        /**
         * @return [char]
         */
        get : function() {
            this.position++;
            return this.array[this.position];
        },

        /**
         * @return [string]
         */
        readLine : function() {
            var sb = "";
            // '\n'が来るまで読み込み
            this.position++;
            for ( ; this.position < this.length; this.position++ ) {
                var c = this.array[this.position];
                if ( c == '\n' ) {
                    break;
                }
                sb += c;
            }
            return sb;
        },

        /**
         * @return [bool]
         */
        ready : function() {
            if ( 0 <= this.position + 1 && this.position + 1 < this.length ) {
                return true;
            } else {
                return false;
            }
        },

        /**
         * @param length [int]
         * @return [void]
         */
        _ensureCapacity : function( _length ) {
            if ( _length > this.array.length ) {
                var add = _length  - this.array.length;
                for( var i = 0; i < add; i++ ){
                    this.array.push( " " );
                }
            }
        },

        /**
         * @param str [string]
         * @return [void]
         */
        write : function( str ) {
            var len = str.length;
            var newSize = this.length + len;
            var offset = length;
            _ensureCapacity( newSize );
            for ( var i = 0; i < len; i++ ) {
                this.array[offset + i] = str.charAt( i );
            }
            this.length = newSize;
        },

        /**
         * @param str [string]
         * @return [void]
         */
        writeLine : function( str ) {
            var len = str.length;
            var newSize = length + len + 1;
            var offset = length;
            _ensureCapacity( newSize );
            for ( var i = 0; i < len; i++ ) {
                this.array[offset + i] = str.charAt( i );
            }
            this.array[offset + len] = '\n';
            this.length = newSize;
        },

        /**
         * @return [void]
         */
        close : function() {
            this.array = null;
            this.length = 0;
        },
    };

}

File: org.kbinani.vsq/TextStream.cs
/*
 * TextMemoryStream.cs
 * Copyright © 2008-2011 kbinani
 *
 * This file is part of org.kbinani.vsq.
 *
 * org.kbinani.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * org.kbinani.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
#if JAVA
package org.kbinani.vsq;

import java.util.*;
import java.io.*;
import org.kbinani.*;
#else
using System;
using System.Text;
using org.kbinani;
using org.kbinani.java.util;
using org.kbinani.java.io;

namespace org.kbinani.vsq
{
    using boolean = System.Boolean;
#endif

#if JAVA
    public class TextStream implements ITextWriter {
#else
    public class TextStream : ITextWriter, IDisposable
    {
#endif
        const int INIT_BUFLEN = 512;

        private char[] array = new char[INIT_BUFLEN];
        private int length = 0;
        private int position = -1;

        public int getPointer()
        {
            return position;
        }

        public void setPointer( int value )
        {
            position = value;
        }

        public char get()
        {
            position++;
            return array[position];
        }

        public String readLine()
        {
            StringBuilder sb = new StringBuilder();
            // '\n'が来るまで読み込み
            position++;
            for ( ; position < length; position++ ) {
                char c = array[position];
                if ( c == '\n' ) {
                    break;
                }
#if JAVA
                sb.append( c );
#else
                sb.Append( c );
#endif
            }
            return sb.ToString();
        }

        public boolean ready()
        {
            if ( 0 <= position + 1 && position + 1 < length ) {
                return true;
            } else {
                return false;
            }
        }

        private void ensureCapacity( int length )
        {
            if ( length > array.Length ) {
                int newLength = length;
                if ( this.length <= 0 ) {
                    newLength = (length * 3) >> 1;
                } else {
                    int order = length / array.Length;
                    if ( order <= 1 ) {
                        order = 2;
                    }
                    newLength = array.Length * order;
                }
#if JAVA
#if JAVA_1_5
                char[] buf = new char[newLength];
                for( int i = 0; i < array.length; i++ ){
                    buf[i] = array[i];
                }
                array = buf;
#else
                array = Arrays.copyOf( array, newLength );
#endif
#else
                Array.Resize( ref array, newLength );
#endif
            }
        }

        public void write( String str )
        {
            int len = PortUtil.getStringLength( str );
            int newSize = length + len;
            int offset = length;
            ensureCapacity( newSize );
            for ( int i = 0; i < len; i++ ) {
#if JAVA
                array[offset + i] = str.charAt( i );
#else
                array[offset + i] = str[i];
#endif
            }
            length = newSize;
        }

        public void writeLine( String str )
        {
            int len = PortUtil.getStringLength( str );
            int newSize = length + len + 1;
            int offset = length;
            ensureCapacity( newSize );
            for ( int i = 0; i < len; i++ ) {
#if JAVA
                array[offset + i] = str.charAt( i );
#else
                array[offset + i] = str[i];
#endif
            }
            array[offset + len] = '\n';
            length = newSize;
        }

        public void close()
        {
            array = null;
            length = 0;
        }

#if !JAVA
        public void Dispose()
        {
            close();
        }
#endif
    }

#if !JAVA
}
#endif
